cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/GitPATAuthenticationStrategy.java

```diff
@@ -16,29 +16,90 @@
 
 package io.cdap.cdap.sourcecontrol;
 
+import com.google.common.base.Throwables;
 import io.cdap.cdap.api.security.store.SecureStore;
+import io.cdap.cdap.api.security.store.SecureStoreData;
 import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+import org.eclipse.jgit.errors.UnsupportedCredentialItem;
+import org.eclipse.jgit.transport.CredentialItem;
 import org.eclipse.jgit.transport.CredentialsProvider;
+import org.eclipse.jgit.transport.URIish;
 import org.eclipse.jgit.transport.UsernamePasswordCredentialsProvider;
 
+import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 
 /**
  * An {@link AuthenticationStrategy} to use with GitHub and Personal Access Tokens.
  */
 public class GitPATAuthenticationStrategy implements AuthenticationStrategy {
   private static final String GITHUB_PAT_USERNAME = "oauth2";
+  private final SecureStorePasswordProvider credentialsProvider;
+
+  public GitPATAuthenticationStrategy(SecureStore secureStore, RepositoryConfig config, String namespaceId) {
+    this.credentialsProvider =
+      new SecureStorePasswordProvider(secureStore, GITHUB_PAT_USERNAME, config.getAuth().getTokenName(), namespaceId);
+  }
+
+  @Override
+  public CredentialsProvider getCredentialsProvider() {
+    return credentialsProvider;
+  }
 
   @Override
-  public CredentialsProvider getCredentialsProvider(SecureStore store,
-                                                    RepositoryConfig config,
-                                                    String namespaceId) throws AuthenticationConfigException {
-    try {
-      byte[] bytes = store.get(namespaceId, config.getAuth().getTokenName()).get();
-      String token = new String(bytes, StandardCharsets.UTF_8);
-      return new UsernamePasswordCredentialsProvider(GITHUB_PAT_USERNAME, token);
-    } catch (Exception e) {
-      throw new AuthenticationConfigException("Failed to get auth token from secure store", e);
+  public void refresh() throws AuthenticationConfigException, IOException {
+    credentialsProvider.refresh();
```

Comment:
this is not ideal because it means the state of the CredentialsProvider is tied to what happens with the AuthenticationStrategy. It will be easier to have bugs where the CredentialsProvider sometimes works and sometimes doesn't.

It would be better to introduce an abstract RefreshableCredentialsProvider class that extends CredentialsProvider and has an abstract refresh() method. Then change getCredentialsProvider to return that.

If that doesn't work for some reason, then it would be better to look up the password and create a new UsernamePasswordCredentialsProvider in each refresh() call, rather than re-use the same CredentialsProvider. This makes the behavior more predictable, either the CredentialsProvider will always work or it will never work.

Reply:
Makes sense. Introduced a RefreshableCredentialsProvider abstract class as suggested.